### PR TITLE
[SYCL] Enable memcpy in libdevice

### PR DESCRIPTION
**Pull request by jinge90**
Signed-off-by: gejin <ge.jin@intel.com>

Currently, memcpy is supported in SYCL device code if "-fno-builtin" is not added. Compiler will convert it to "llvm.memcpy..." intrinsic and llvm-spirv will convert it to OpCopyMemorySized for jit.

However, if "-fno-builtin" is added, compiler will call libc function "memcpy" directly and it requires libdevice support. This PR support memcpy in libdevice, so developers can use it in SYCL device code no matter whether "-fno-builtin" is added or not.

### CODE REVIEW

File: libdevice/fallback-cstring.cpp [hunk 1]
```diff
@@ -0,0 +1,15 @@
+//==--- fallback-cassert.cpp - device agnostic implementation of C assert --==//
```
Comment (kbobrovs) [inline libdevice/fallback-cstring.cpp:1]: wrong title

File: clang/test/Driver/sycl-device-lib.cpp [hunk 2]
```diff
@@ -21,6 +21,7 @@
 // SYCL_DEVICE_LIB_UNBUNDLE_DEFAULT-NEXT: clang-offload-bundler{{.*}} "-type=o" "-targets=sycl-spir64-unknown-unknown-sycldevice" "-inputs={{.*}}libsycl-cmath.o" "-outputs={{.*}}libsycl-cmath-{{.*}}.o" "-unbundle"
 // SYCL_DEVICE_LIB_UNBUNDLE_DEFAULT-NEXT: clang-offload-bundler{{.*}} "-type=o" "-targets=sycl-spir64-unknown-unknown-sycldevice" "-inputs={{.*}}libsycl-cmath-fp64.o" "-outputs={{.*}}libsycl-cmath-fp64-{{.*}}.o" "-unbundle"
 // SYCL_DEVICE_LIB_UNBUNDLE_DEFAULT-NEXT: clang-offload-bundler{{.*}} "-type=o" "-targets=sycl-spir64-unknown-unknown-sycldevice" "-inputs={{.*}}libsycl-fallback-cassert.o" "-outputs={{.*}}libsycl-fallback-cassert-{{.*}}.o" "-unbundle"
+// SYCL_DEVICE_LIB_UNBUNDLE_DEFAULT-NEXT: clang-offload-bundler{{.*}} "-type=o" "-targets=sycl-spir64-unknown-unknown-sycldevice" "-inputs={{.*}}libsycl-fallback-cstring.o" "-outputs={{.*}}libsycl-fallback-cstring-{{.*}}.o" "-unbundle"
```
Comment (bader) [inline clang/test/Driver/sycl-device-lib.cpp:24]: Please, add clang/test/Driver/Inputs/SYCL/lib/libsycl-fallback-cstring.o. The fact that the test is passing w/o this file proves that https://github.com/intel/llvm/issues/3630 is not fixed. I think `--sysroot` just **adds** new search paths, whereas we need to **replace** default paths with the new location.

File: libdevice/cmake/modules/SYCLLibdevice.cmake [hunk 3]
```diff
@@ -86,6 +86,15 @@ add_custom_command(OUTPUT ${spv_binary_dir}/libsycl-fallback-cassert.spv
                    DEPENDS wrapper.h device.h clang spirv_vars.h llvm-spirv
                    VERBATIM)
 
+add_custom_command(OUTPUT ${spv_binary_dir}/libsycl-fallback-cstring.spv
+                   COMMAND ${clang} -S -fsycl-device-only -fno-sycl-use-bitcode
```
Comment (bader) [inline libdevice/cmake/modules/SYCLLibdevice.cmake:90]: What is `-S` for?

**Review by bader — DISMISSED** [review]

File: clang/test/Driver/sycl-device-lib.cpp (hunk 2 above)
Comment (jinge90) [reply clang/test/Driver/sycl-device-lib.cpp:24]: Hi, @bader 
Currently, we follow CUDA( https://github.com/intel/llvm/blob/sycl/clang/lib/Driver/ToolChains/Cuda.cpp#L168 ), sysroot will be added as a candidate and we will go through all candidates to find existing libdevice binary. If we added fake binary for all devicelib, the clang lit test should pass even if we didn't buid libdevice.

Comment (bader) [reply clang/test/Driver/sycl-device-lib.cpp:24]: There is a difference between SYCL and CUDA. SYCL toolchain always adds driver directory as a candidate, whereas CUDA configures candidates via `sysroot` only.

Comment (bader) [reply clang/test/Driver/sycl-device-lib.cpp:24]: I suggest we change [this line](https://github.com/intel/llvm/blob/sycl/clang/lib/Driver/ToolChains/SYCL.cpp#L27): from
```c++
InstallationCandidates.emplace_back(D.Dir + "/..");
```
to
```c++
InstallationCandidates.emplace_back(D.SysRoot + D.Dir + "/..");
```

**Review by romanovvlad — APPROVED** [review]
sycl/include and sycl/source changes  LGTM

**Review by Pennycook — APPROVED** [review]

**Review by AlexeySachkov — DISMISSED** [review]
`sycl-post-link` changes LGTM

**Review by vzakhari — DISMISSED** [review]
`libdevice` changes look good to me.

File: clang/test/Driver/sycl-device-lib.cpp (hunk 2 above)
Comment (jinge90) [reply clang/test/Driver/sycl-device-lib.cpp:24]: Hi, @bader 
Currently, the D.dir is an absolute path and is it correct for us to concatenate D.SysRoot with an absolute path?
Hi, @mdtoguchi , do you have any comments?
Thanks very much.

Comment (bader) [reply clang/test/Driver/sycl-device-lib.cpp:24]: > Currently, the D.dir is an absolute path and is it correct for us to concatenate D.SysRoot with an absolute path?

Yes.

**Review by romanovvlad — APPROVED** [review]
RT part LGTM

File: clang/test/Driver/sycl-device-lib.cpp (hunk 2 above)
Comment (bader) [reply clang/test/Driver/sycl-device-lib.cpp:24]: Marking this conversation as resolved:
- `clang/test/Driver/Inputs/SYCL/lib/libsycl-fallback-cstring.o` is added
- https://github.com/intel/llvm/issues/3630 can be handled by another PR.

**Comment by bader** [comment]
@AGindinson, @mdtoguchi, ping.

**Review by vzakhari — APPROVED** [review]

**Review by mdtoguchi — APPROVED** [review]

**Comment by jinge90** [comment]
Hi, @bader and @romanovvlad 
Could you help check if this patch can be merged now?
Thanks very much.

**Comment by romanovvlad** [comment]
> 
> 
> Hi, @bader and @romanovvlad
> Could you help check if this patch can be merged now?
> Thanks very much.

No, it cannot be.
@mlychkov  @AGindinson  Could you please review?

**Review by mlychkov — APPROVED** [review]

**Pull request merged**